### PR TITLE
Remove `log4j2.enable.threadlocals` property

### PR DESCRIPTION
**Pull request by vy**
This PR addresses #2105 and removes `log4j2.enable.threadlocals` property. While doing so, following major changes are also implemented:

- `ThreadLocal`-based recycler is removed
- `ParametrizedMessageFactory` is removed – switched to `ReusableMessageFactory` everywhere
- `ThreadNameCachingStrategy` is removed – Java 17 renders it redundant
- `DefaultThreadContextMap` is removed – Other implementations serve fine with recyclers, which don't have any `ThreadLocal` issues anymore. Weirdly enough, `DefaultThreadContextMap`, which is _the_ fallback used when thread locals are disabled, was using thread locals too.

**Question:** Similar to `ParametrizedMessageFactory`, I doubt if `DefaultLogEventFactory` makes sense anymore. Can we make the `ReusableLogEventFactory` default instead?

### CODE REVIEW

**Review by jvz — APPROVED** [review]
Amazing diff!

**Review by ppkarwasz — COMMENTED** [review]
There is something I don't understand about this PR: we **do** want a `ThreadLocalRecyclerFactoryProvider` out-of-the-box.

But we want to be enabled/disabled using the `Recycler.factory = "threadLocal"` and `Recycler.factory = "dummy"` properties.

It's great you removed all the trailing `ThreadLocal` instances, but I think you should restore the `ThreadLocalRecyclerFactory`.

I am not sure what recycler factory should be the default one though: Spring Boot applications usually contain a Servlet API, but have only a single app running, so a default of `threadLocal` sounds good to me. If there is a way to detect the usage of virtual threads, we could switch the default to "dummy".

File: log4j-api/src/main/java/org/apache/logging/log4j/message/MessageFactory.java [hunk 1]
```diff
@@ -58,7 +57,6 @@ default Message newMessage(final String message) {
      * @param params
      *            the message parameters
      * @return a new message
-     * @see ParameterizedMessageFactory
      * @see StringFormatterMessageFactory
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-api/src/main/java/org/apache/logging/log4j/message/MessageFactory.java]: Fixed in 26a00aab4e6b03cf2087d2eddc9c6057ff6f6a5f.

**Comment by vy** [comment]
@ppkarwasz, I don't think you want a thread local recycler. :stuck_out_tongue_winking_eye: In the face of virtual threads, thread locals are a risky liability. As a matter of fact, they already were actually; hence, several _"if not running in a web application"_, _"if only using JDK classes"_, etc. guards surrounding them. The new default, the queueing recycler, is almost as good as the thread local recycler. Those who need more efficiency, can use `log4j-jctools` too. All in all, no need for thread locals anymore really.

**Comment by vy** [comment]
@ppkarwasz, I placed the thread local recycler back. I want to move on with this PR.

**Comment by rgoers** [comment]
I didn't see any of the doc files listed in the PR. If the property was removed it should be removed from the documentation as well.

**Comment by vy** [comment]
> I didn't see any of the doc files listed in the PR. If the property was removed it should be removed from the documentation as well.

Fixed in e6044c995721e6ab0cc207c7536ee37c9a638323.

**Pull request merged**